From 2fdc268189e4cf6bf3ac82b370c0bff7b9f7c7f4 Mon Sep 17 00:00:00 2001
From: Douglas Reid <douglas-reid@users.noreply.github.com>
Date: Fri, 4 Mar 2022 11:04:43 -0800
Subject: TelemetryAPI: Disable Stackdriver logging if not configured (by
 default, or otherwise)  (#37708)

* TelemetryAPI: Disable Stackdriver logging if not configured

* lint and release notes
---
 pilot/pkg/model/telemetry.go      | 4 ++++
 pilot/pkg/model/telemetry_test.go | 6 +++---
 releasenotes/notes/37708.yaml     | 8 ++++++++
 3 files changed, 15 insertions(+), 3 deletions(-)
 create mode 100644 releasenotes/notes/37708.yaml

diff --git a/pilot/pkg/model/telemetry.go b/pilot/pkg/model/telemetry.go
index 8efbd5d3e2..1bda499fcc 100644
--- a/pilot/pkg/model/telemetry.go
+++ b/pilot/pkg/model/telemetry.go
@@ -855,7 +855,11 @@ func generateSDConfig(class networking.ListenerClass, telemetryConfig telemetryF
 				cfg.AccessLogging = sd.PluginConfig_ERRORS_ONLY
 			}
 		}
+	} else {
+		// The field is deprecated, but until it is removed we need to set it.
+		cfg.DisableServerAccessLogging = true // nolint: staticcheck
 	}
+
 	cfg.MetricExpiryDuration = durationpb.New(1 * time.Hour)
 	// In WASM we are not actually processing protobuf at all, so we need to encode this to JSON
 	cfgJSON, _ := protomarshal.MarshalProtoNames(&cfg)
diff --git a/pilot/pkg/model/telemetry_test.go b/pilot/pkg/model/telemetry_test.go
index 0ebfe4e92b..77d1e267b2 100644
--- a/pilot/pkg/model/telemetry_test.go
+++ b/pilot/pkg/model/telemetry_test.go
@@ -687,7 +687,7 @@ func TestTelemetryFilters(t *testing.T) {
 			networking.ListenerProtocolHTTP,
 			nil,
 			map[string]string{
-				"istio.stackdriver": `{"metric_expiry_duration":"3600s"}`,
+				"istio.stackdriver": `{"disable_server_access_logging":true,"metric_expiry_duration":"3600s"}`,
 			},
 		},
 		{
@@ -713,7 +713,7 @@ func TestTelemetryFilters(t *testing.T) {
 			networking.ListenerProtocolHTTP,
 			nil,
 			map[string]string{
-				"istio.stackdriver": `{"metric_expiry_duration":"3600s"}`,
+				"istio.stackdriver": `{"disable_server_access_logging":true,"metric_expiry_duration":"3600s"}`,
 			},
 		},
 		{
@@ -753,7 +753,7 @@ func TestTelemetryFilters(t *testing.T) {
 			networking.ListenerProtocolHTTP,
 			&meshconfig.MeshConfig_DefaultProviders{Metrics: []string{"prometheus"}},
 			map[string]string{
-				"istio.stackdriver": `{"metric_expiry_duration":"3600s"}`,
+				"istio.stackdriver": `{"disable_server_access_logging":true,"metric_expiry_duration":"3600s"}`,
 			},
 		},
 		{
diff --git a/releasenotes/notes/37708.yaml b/releasenotes/notes/37708.yaml
new file mode 100644
index 0000000000..956671b75c
--- /dev/null
+++ b/releasenotes/notes/37708.yaml
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: telemetry
+issue:
+  - https://github.com/istio/istio/issues/37667
+releaseNotes:
+  - |
+    **Fixed** an issue where when enabling Stackdriver metrics collection with the Telemetry API, logging was incorrectly enabled in certain scenarios.
-- 
2.35.3

